### ISSUE TITLE
4.16 support :)

### ISSUE THREAD

**Issue by Youdaman** [ncsoft/Unreal.js#170]
Tried the manual install with a new 4.16 project but the build fails.

```
1>------ Rebuild All started: Project: MyProject, Configuration: Development_Editor x64 ------
1>  Cleaning MyProjectEditor Binaries...
1>  Messages while compiling C:\Users\Youdaman\Desktop\unreal\MyProject\Intermediate\Build\BuildRules\MyProjectModuleRules.dll:
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptConsole\JavascriptConsole.Build.cs(7,12): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptEditor\JavascriptEditor.Build.cs(24,12): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptGraphEditor\JavascriptGraphEditor.Build.cs(7,10): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptHttp\JavascriptHttp.Build.cs(5,9): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptUMG\JavascriptUMG.Build.cs(5,9): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptWebSocket\JavascriptWebSocket.Build.cs(29,12): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptWebSocket\JavascriptWebSocket.Build.cs(46,13): error CS1502: The best overloaded method match for 'UnrealBuildTool.ModuleRules.AddEngineThirdPartyPrivateStaticDependencies(UnrealBuildTool.ReadOnlyTargetRules, params string[])' has some invalid arguments
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptWebSocket\JavascriptWebSocket.Build.cs(46,58): error CS1503: Argument 1: cannot convert from 'UnrealBuildTool.TargetInfo' to 'UnrealBuildTool.ReadOnlyTargetRules'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptWebSocket\JavascriptWebSocket.Build.cs(46,65): error CS1503: Argument 2: cannot convert from 'string' to 'string[]'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\V8\V8.Build.cs(36,12): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>EXEC : error : UnrealBuildTool encountered an error while compiling source files
1>  Creating makefile for MyProjectEditor (no existing makefile)
1>  Messages while compiling C:\Users\Youdaman\Desktop\unreal\MyProject\Intermediate\Build\BuildRules\MyProjectModuleRules.dll:
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptConsole\JavascriptConsole.Build.cs(7,12): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptEditor\JavascriptEditor.Build.cs(24,12): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptGraphEditor\JavascriptGraphEditor.Build.cs(7,10): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptHttp\JavascriptHttp.Build.cs(5,9): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptUMG\JavascriptUMG.Build.cs(5,9): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptWebSocket\JavascriptWebSocket.Build.cs(29,12): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptWebSocket\JavascriptWebSocket.Build.cs(46,13): error CS1502: The best overloaded method match for 'UnrealBuildTool.ModuleRules.AddEngineThirdPartyPrivateStaticDependencies(UnrealBuildTool.ReadOnlyTargetRules, params string[])' has some invalid arguments
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptWebSocket\JavascriptWebSocket.Build.cs(46,58): error CS1503: Argument 1: cannot convert from 'UnrealBuildTool.TargetInfo' to 'UnrealBuildTool.ReadOnlyTargetRules'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\JavascriptWebSocket\JavascriptWebSocket.Build.cs(46,65): error CS1503: Argument 2: cannot convert from 'string' to 'string[]'
1>c:\Users\Youdaman\Desktop\unreal\MyProject\Plugins\UnrealJS\Source\V8\V8.Build.cs(36,12): warning CS0618: 'UnrealBuildTool.ModuleRules.ModuleRules()' is obsolete: 'Please change your module constructor to take a ReadOnlyTargetRules parameter, and pass it to the base class constructor (eg. "MyModuleRules(ReadOnlyTargetRules Target) : base(Target)").'
1>EXEC : error : UnrealBuildTool encountered an error while compiling source files
1>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.MakeFile.Targets(46,5): error MSB3073: The command ""C:\Program Files\Epic Games\UE_4.16\Engine\Build\BatchFiles\Rebuild.bat" MyProjectEditor Win64 Development "C:\Users\Youdaman\Desktop\unreal\MyProject\MyProject.uproject" -waitmutex" exited with code -1.
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
```

**Comment by crocuis** [comment]
It will be updated after 4.16 released.

**Comment by crocuis** [comment]
@Youdaman 4.16 supported now!

**Comment by FelschR** [comment]
@crocuis It seems the update wasn't released in the marketplace yet?

**Comment by andrewmunro** [comment]
@crocuis Can you re-publish to https://www.unrealengine.com/marketplace/unrealjs and close this issue?

**Comment by crocuis** [comment]
It needs more time :(

**Comment by Youdaman** [comment]
@crocuis there was also a delay with 4.15 and we (@nakosung and myself) let Alexander from Epic know. Not sure why they're taking so long? There seems to be others also upset about the marketplace: https://twitter.com/Luos_83/status/875048304253558785

I would think that because it's NCSoft that Epic would at least give it more priority :P

**Comment by natehitchcock** [comment]
Is Unreal JS compatible with 4.16 yet? I'm still getting a ton of build errors with the manual install

**Comment by Youdaman** [comment]
@natehitchcock I just tried the manual install with a fresh blank 4.16.1 C++ project and it worked fine.

**Comment by natehitchcock** [comment]
I'm getting some very strange results from building.
Clean project + original unreal.js => builds fine
Our project + original unreal.js => builds fine
Our project + our fork of unreal.js => everything explodes
Our project + original unreal.js with our changes hand copied over => builds fine

Our fork is up-to-date with the original repo.
Diffing the two repos, they are identical except for a couple small changes (our changes consist mostly of modifying the typings generator and adding some functions to the javascript function library).

We haven't added or removed any header dependencies, but all the errors I'm getting are about missing UE4 classes, and Unreal header files not found. It's almost like the .pch is not being compiled in properly, except all the build.cs files are identical and set up correctly.

[Here is a link to our fork of Unreal.js](https://github.com/SoulboundStudios/Unreal.js-core)

**Comment by crocuis** [comment]
Show me your build error message plz :)

**Comment by natehitchcock** [comment]
I figured out what was causing our compile errors. The errors were about missing V8 and unreal types, and that caused a whole avalanche of related errors. All of these types are linked by injecting the .pch into each translation unit (blob? discussed [here](https://docs.unrealengine.com/latest/INT/Programming/UnrealBuildSystem/Configuration/)). 

That forced injection of the .pch wasn't happening for us.

We recently moved a lot of our source code over from perforce to git, so all the .cpp files (that we hadn't modified since) had the 'read-only' flag set, which told the UBT to not bother re-injecting the .pch into each file (even though we were cleaning and building- I'm not sure if that is a bug in UBT, or just our process).

Un-setting the read-only flag allow the .pch to be injected into each file again, which fixed all of our errors.

**TL;DR: Read-only flag was checked by accident, had bad consequences in the unreal build process**
